### PR TITLE
`ultralytics 8.0.134` add MobileSAM support

### PR DESCRIPTION
**Pull request by ChaoningZhang**
related issue: https://github.com/ultralytics/ultralytics/pull/3474

I have read the CLA Document and I sign the CLA

### Summary
📱🎨🧠

This pull request adds a new module `ultralytics/vit/mobilesam` that implements the MobileSAM model, a self-attention and mobile network for object segmentation. It also updates the `ultralytics` package and the documentation to support the new model. The MobileSAM model uses a vision transformer (ViT) backbone and a transformer decoder to predict masks from an image and input prompts. The module provides various classes and functions for building, loading, predicting, and evaluating the model, as well as utilities for mask generation and resizing.

> _`MobileSAM` joins us_
> _Self-attention for images_
> _Autumn of objects_

### Walkthrough
*  Add the MobileSAM model, a novel method for object segmentation using point prompts and a transformer-based architecture ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-51d382352f210bfa1b95b0a2c342134acd6912ce7da775c0c34b5948d0bb3863R1-R90), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceR7), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceL13-R14), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-5597b74444fe27a26922336224a04bd9ed28414e3b2d8eb28dc09f68cf9e367dL5-R6), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-268a62f5b4d1b1ba9c323a1ed7a2ab1d073ff4ac720c5fbab34e7ff05fa8da22R1-R6), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-468c647637087290ae6e601d47984f96a61c425918585339fa4d495dfc530247R1-R311), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-a714f56000e9d8c01ac1dfbe153d2fae133b02d21f04ee374885eb09996a3b3eR1-R94), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-117fc0c2a2580afc91f130ff1186257b2d95fd42ef21c7d5c87e260fd21e7affL1-R130), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-2917b52cecbe928a46e6b781eb7da764524f54826b2d0be52aec18d9ba4ae57dR1-R87), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-5185c7280e8c41e7e0ba09db906dd36c494e03a8ec1725566bdb306df51284bcR1), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-82c4209a639253b10bed575463dce192d430395c8909d819b1a1e0331bcba50eR1-R159), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-7044b12ff3fb8cf900f54f886724a231e48727f32c907d91d0f514daf08a9ec7R1-R353), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-74661277dd0f2614bedb1d22aafc53b24f719384cd4b41a83144c6fb172a0663R1-R242), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-d61b012423be55ac8ef371481f48fd6a8a85a5043770ac13f35f9eb3070bb00eR1-R173), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-3891442902adbca7befdedb7c4cabeca9912e57a9369e8f36e2e466155302977R1-R235), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b9b212f9f0d92979b6199adcfb194ac951c1bb68a97978f0d6eedf1be17fd9b8L1-R133), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-ff51b2d28d911ac05d1a88597018be80aac41d670db9eb719bcf9c2c1633d2c9L1-R49))
* Import the MobileSAM class from the `mobilesam` module in the `ultralytics` and `vit` packages, and add it to the `__all__` list in the `ultralytics` package, to make it accessible to users ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceR7), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceL13-R14), [link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-5597b74444fe27a26922336224a04bd9ed28414e3b2d8eb28dc09f68cf9e367dL5-R6))
* Define the `MobileSAM` class in the `model.py` file, which inherits from the `BasePredictor` class and provides methods for initializing, predicting, training, validating, and getting information about the MobileSAM model ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-2917b52cecbe928a46e6b781eb7da764524f54826b2d0be52aec18d9ba4ae57dR1-R87))
* Define the `Sam` class in the `sam.py` file, which inherits from `nn.Module` and implements the SAM model that predicts object masks from an image and input prompts ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-d61b012423be55ac8ef371481f48fd6a8a85a5043770ac13f35f9eb3070bb00eR1-R173))
* Define the `TwoWayTransformer` class, the `TwoWayAttentionBlock` class, and the `Attention` class in the `transformer.py` file, which are used by the SAM model to implement the transformer decoder that attends to an input image using queries whose positional embedding is supplied ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-3891442902adbca7befdedb7c4cabeca9912e57a9369e8f36e2e466155302977R1-R235))
* Define the `MaskDecoder` class and the `MLP` class in the `decoders.py` file, which are used by the SAM model to predict masks from the image and prompt embeddings, using a transformer architecture and a hypernetwork ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-82c4209a639253b10bed575463dce192d430395c8909d819b1a1e0331bcba50eR1-R159))
* Define the `SamAutomaticMaskGenerator` class in the `mask_generator.py` file, which uses a SAM model to generate masks for the entire image, given a grid of point prompts ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-7044b12ff3fb8cf900f54f886724a231e48727f32c907d91d0f514daf08a9ec7R1-R353))
* Define the `PromptPredictor` class in the `prompt_predictor.py` file, which uses SAM to calculate the image embedding for an image, and then allows repeated, efficient mask prediction given prompts ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-74661277dd0f2614bedb1d22aafc53b24f719384cd4b41a83144c6fb172a0663R1-R242))
* Define the `Predictor` class in the `predict.py` file, which inherits from the `BasePredictor` class and provides methods for preprocessing, postprocessing, and predicting masks using the MobileSAM model ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-b9b212f9f0d92979b6199adcfb194ac951c1bb68a97978f0d6eedf1be17fd9b8L1-R133))
* Define the `ResizeLongestSide` class in the `autosize.py` file, which resizes images and coordinates to the longest side target length, and provides methods for transforming numpy arrays and torch tensors ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-a714f56000e9d8c01ac1dfbe153d2fae133b02d21f04ee374885eb09996a3b3eR1-R94))
* Define various utility functions for mask generation, such as cropping, filtering, encoding, and postprocessing masks, in the `amg.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-468c647637087290ae6e601d47984f96a61c425918585339fa4d495dfc530247R1-R311))
* Define functions for building different SAM models with different ViT backbones, such as ViT-H, ViT-L, ViT-B, and Tiny-ViT, in the `build.py` file, and a `sam_model_map` dictionary that maps the model checkpoints to the corresponding model builders ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-117fc0c2a2580afc91f130ff1186257b2d95fd42ef21c7d5c87e260fd21e7affL1-R130))
* Define a `setup_model` function in the `setup_mobile_sam.py` file, which loads the Tiny-ViT backbone and the other components of the MobileSAM model from the given checkpoint ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-ff51b2d28d911ac05d1a88597018be80aac41d670db9eb719bcf9c2c1633d2c9L1-R49))
* Add the metadata and the introduction for the MobileSAM model in the `docs/models/mobilesam.md` file, which describes the main features, advantages, and usage of MobileSAM, as well as provides links to the paper, demo, and bibtex ([link](https://github.com/ultralytics/ultralytics/pull/3474/files?diff=unified&w=0#diff-51d382352f210bfa1b95b0a2c342134acd6912ce7da775c0c34b5948d0bb3863R1-R90))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Ultralytics has introduced MobileSAM, a light-weight model optimized for mobile applications, offering speed and full functionality for computer vision tasks on CPU-only devices.

### 📊 Key Changes
- Added `MobileSAM` documentation to the models index.
- Created a new documentation page `mobile-sam.md` for `MobileSAM`, explaining its purpose, usage, and performance details.

### 🎯 Purpose & Impact
- The introduction of **MobileSAM** aims to bring the powerful segmentation capability of SAM to mobile platforms, where computational resources are limited.
- **Users can expect** noticeable performance improvements when performing image segmentation tasks on mobile or edge devices, with an emphasis on real-world usability and integration into mobile applications.
- **Developers can utilize** `MobileSAM` for various projects that require efficient image segmentation without compromising on accuracy or speed.

### CODE REVIEW

**Review by ChaoningZhang — COMMENTED** [review]
I have read the CLA Document and I sign the CLA

**Review by ChaoningZhang — COMMENTED** [review]
I have read the CLA Document and I sign the CLA

**Comment by ChaoningZhang** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I sign the CLA
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

I have read the CLA Document and I sign the CLA

**Comment by ChaoningZhang** [comment]
I have read the CLA Document and I sign the CLA

File: picture1.jpg
Comment (glenn-jocher) [inline picture1.jpg:1]: Can we delete this picture1.jpg file? We have default images available in ultralytics/assets/, I've updated Doc page to use ultralytics/assets/bus.jpg.

**Comment by glenn-jocher** [comment]
@ChaoningZhang can you please delete picture1.png from this PR? I've already updated the docs to use the existing bus.jpg asset.

**Comment by Laughing-q** [comment]
@glenn-jocher Actually I updated the pic to ultralytics/assets/zidane.jpg😂 and also updated the bboxes and points in docs. Currently the example looks like:
```python
from ultralytics import SAM

# Load a model
model = SAM('mobile_sam.pt')
# Run inference with bboxes prompt
model('ultralytics/assets/zidane.jpg', bboxes=[439, 437, 524, 709])
```
zidane

```python
# Run inference with points prompt
model.predict('ultralytics/assets/zidane.jpg', points=[900, 370], labels=[1])
```
zidane

**Comment by glenn-jocher** [comment]
@Laughing-q oh perfect! I just deleted picture1.png programmatically now. Let me know when the PR is ready and I'll review and merge it.

**Comment by Laughing-q** [comment]
@glenn-jocher @AyushExel @ChaoningZhang I've managed to update and rebuild SAM-related code. Everything works properly now, and moiblesam model can be seamlessly used with SAM.
You can create both SAM and MoibleSAM models within the same way:
```python
from ultralytics import SAM
model = SAM("mobile_sam.pt")
```
I added `bboxes` and `points` as args for prompt inference, passing `bboxes` or `points` for prompt inference. 
```python
# running prompt inference
model('ultralytics/assets/zidane.jpg', bboxes=[439, 437, 524, 709])
model.predict('ultralytics/assets/zidane.jpg', points=[900, 370], labels=[1])
```

It'll segment the whole image if you don't pass any prompt-related args(bboxes/labels/points/masks). And it'll also return the corresponding predicted bboxes then we can handle these predictions easily with our Results class just like yolo models.
```python
# segment the whole image
model.predict('ultralytics/assets/zidane.jpg')
```
img_v2_f9e40b9a-6fb8-487a-91ce-3cf81b010bdg

@glenn-jocher  Basically this PR is all set, except some functions in ultralytics/vit/sam/amg.py used by sam/predict.py still have space to optimize. Also please update the docs if you find anything doesn't seem right since somehow I'm still not able to build mkdocs.

**Comment by Laughing-q** [comment]
@glenn-jocher ok I think this PR is ready. 
BTW I noticed you removed `mobilesam.md` file but you also rename `mobilesam.md` to `mobile-sam.md` in the index.md file so I added `mobile-sam.md` file back here. https://github.com/ultralytics/ultralytics/pull/3474/commits/96ff488a6f2e980119f78055582cb467fddf068f

**Comment by glenn-jocher** [comment]
@Laughing-q ok got it! I'll review everything!

**Comment by Laughing-q** [comment]
@glenn-jocher oh right, our internal resize operation `LetterBox` is to put image in the middle and pad top/bottom/left/right sides, but preprocessing of sam models needs to put image at top-left so I added an additional arg `center` in `LetterBox` to implement it. 
Currently with `center=True` which is default, LetterBox works like before, but with `center=False` it would put the image at top-left and only pad the right-bottom. 
Same reason for the additional arg in `scale_masks/scale_boxes/scale_coords`.

**Comment by glenn-jocher** [comment]
@ChaoningZhang @Laughing-q PR merged!!! Thank you for your contributions and all the hard work!
